src/http-client/node-http2-client.ts

```diff
@@ -0,0 +1,222 @@
+import http2 from "http2";
+
+import { HTTPClient, HTTPRequest, HTTPResponse } from "./index";
+import { NetworkError, ProtocolError } from "../errors";
+import { QueryRequest } from "../wire-protocol";
+
+const { HTTP2_HEADER_PATH, HTTP2_HEADER_METHOD, HTTP2_HEADER_STATUS } =
+  http2.constants;
+
+/**
+ * An implementation for {@link HTTPClient} that uses the node http package
+ */
+export class NodeHTTP2Client implements HTTPClient {
+  static #client: NodeHTTP2Client | null = null;
+  #sessionMap: Map<string, SessionWrapper> = new Map();
+
+  /**
+   * @remarks Private constructor means you must instantiate with
+   * {@link NodeHTTP2Client.getClient}
+   */
+  // eslint-disable-next-line @typescript-eslint/no-empty-function
+  private constructor() {}
+
+  static getClient() {
+    if (this.#client) {
+      return this.#client;
+    }
+
+    this.#client = new NodeHTTP2Client();
+    return this.#client;
+  }
+
+  /** {@inheritDoc HTTPClient.request} */
+  async request(httpRequest: HTTPRequest): Promise<HTTPResponse> {
+    const session = this.#getSession(httpRequest.url);
```

Comment:
Let's take a P2 or P3 to have the maxConns of the ClientConfiguration control how many streams we open per session: https://nodejs.org/api/http2.html#collecting-http2-performance-metrics

There's an argument there that some clients may want to use that as a throttle on how much work their client is doing.

#82 fits well with it because we should rename max conns to be something like "max streams"

Let's sync with the go and python driver crew to see what they are doing here as well. Not blocking for P1

Reply:
I created [FE-3290](https://faunadb.atlassian.net/browse/FE-3290) to track outside of P1 fit and finish